jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java

```diff
@@ -103,6 +146,111 @@ public String toString()
         }
     }
 
+    /**
+     * Visitor
+     *
+     * A FileVisitor for walking a subtree of paths. The Scanner uses
+     * this to examine the dirs and files it has been asked to scan.
+     */
+    public class Visitor implements FileVisitor<Path>
+    {
+        Map<String, TimeNSize> scanInfoMap;
+        
+        public Visitor(Map<String, TimeNSize> scanInfoMap)
+        {
+            this.scanInfoMap = scanInfoMap;
+        }
+        
+        @Override
+        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException
+        {
+            File f = dir.toFile();
+            
+            if (!f.exists())
+                return FileVisitResult.SKIP_SUBTREE;
+
+            if (_reportDirs)
+            {
+                boolean accepted = false;
+                IncludeExcludeSet<PathMatcher, Path> includesExcludes = _scannables.get(dir);
+                if (includesExcludes != null)
+                { 
+                    //accepted if not explicitly excluded and either is explicitly included or there are no explicit inclusions
+                    Boolean result = includesExcludes.isIncludedAndNotExcluded(dir);
+                    if (Boolean.TRUE == result)
+                        accepted = true;
+                    else if (result == null && !includesExcludes.hasIncludes())
+                        accepted = true;
+                }
+                else if (_filter == null || _filter.accept(f.getParentFile(), f.getName()))
+                    accepted = true;
+                
+                if (accepted)
+                {
+                    scanInfoMap.put(f.getCanonicalPath(), new TimeNSize(f.lastModified(), f.isDirectory() ? 0 : f.length()));
+                    if (LOG.isDebugEnabled()) LOG.debug("scan accepted dir {} mod={}", f, f.lastModified());
+                }
+            }
+            
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException
+        {
+            File f = file.toFile();
+            
+            if (!f.exists())
+                return FileVisitResult.CONTINUE;
+            
+            if (f.isDirectory())
+                return FileVisitResult.CONTINUE; //handled by preVisitDirectory
+
+            boolean accepted = false;
+            Path tmp = file.toRealPath();
+            IncludeExcludeSet<PathMatcher, Path> includesExcludes = null;
+            includesExcludes = _scannables.get(tmp);
+            while (includesExcludes == null && tmp.getNameCount() > 0)
+            {
+                tmp = tmp.getParent();
+                includesExcludes = _scannables.get(tmp);
+            }
```

Comment:
This logic worries me on two fronts.... firstly things like symlinks are just not going to work - they need to be top down.
Secondly, I don't really understand what it would mean to have different include and excludes sets at /foo, /foo/bar and /foo/bar/bob!

I think there should just be an IncludeExclude with every directory root configured and that is passwed down somehow during the visit, rather than rediscovered all the time.

If somebody watches /dir and /dir/subdir then /dir/subdir/subsub is walked twice, once with the includeExcludeSet of /dir and another time with  the includeExcludeSet of /dir/subdir

Reply:
Done. Visitor now takes the root path, and the IncludeExcludeSet of the root path.